File: functions/callNodeFunction/createCertificate/generateAchievementCertificate.js
import got from "got";
import { request, gql } from "graphql-request";
import logger from "../index.js";
import {
  getStorageBucketURL,
  saveCertificateToBucket,
  updateCourseEnrollmentRecord,
} from "./certificateUtils.js";

/**
 * Generates an achievement certificate for a user.
 *
 * @param {Object} courseEnrollment The course enrollment data for the user.
 * @param {string} bucket The name of the Google Cloud bucket to save the certificate to.
 * @returns {Promise<void>} A promise that resolves when the certificate has been generated and saved.
 */
export const generateAchievementCertificate = async (
  courseEnrollment,
  bucket
) => {
  try {
    // Set online_courses and practical_project according to recordType
    const recordType =
      courseEnrollment.User.AchievementRecordAuthors[0].AchievementRecord
        .AchievementOption.recordType;
    const recordTitle =
      courseEnrollment.User.AchievementRecordAuthors[0].AchievementRecord
        .AchievementOption.title;
    const online_courses = recordType === "ONLINE_COURSE" ? recordTitle : "";
    const practical_project = recordType === "DOCUMENTATION" ? recordTitle : "";

    // Set certificate text to learning goals or empty string if learning goals is null
    const certificate = (courseEnrollment.Course.learningGoals || "")
      .split("\n")
      .filter((line) => line.trim() !== "")
      .map((line) => `- ${line.trim()}`)
      .join("\n");

    // Get the certificate template URL
    const certificateTemplateURL = await getStorageBucketURL(courseEnrollment.Course.Program.achievementCertificateTemplateURL, bucket)

    // Log the certificate template URL
    logger.debug(
      `Certificate template URL retrieved: ${certificateTemplateURL}`
    );
      
    // Construct the certificateData object for the certificate generation request
    const certificateData = {
      json: {
        template: certificateTemplateURL[0],
        full_name: `${courseEnrollment.User.firstName} ${courseEnrollment.User.lastName}`,
        semester: courseEnrollment.Course.Program.title,
        course_name: courseEnrollment.Course.title,
        ects: courseEnrollment.Course.ects,
        practical_project: practical_project,
        online_courses: online_courses,
        certificate_text: certificate,
      },
    };
    logger.debug(`Certificate data: ${JSON.stringify(certificateData)}`);

    // Send the certificate generation request to the certificate generation service
    const url = "https://edu-old.opencampus.sh/create_certificate_rest";
    const generatedCertificate = await got.post(url, certificateData).json();

    // Log the generated certificate
    logger.debug(
      `Generated achievement certificate for user ${courseEnrollment.User.id} and course ${courseEnrollment.Course.id}`
    );

    // Save the certificate to the Google Cloud Storage bucket
    const path = await saveCertificateToBucket(
      generatedCertificate,
      'achievement',
      courseEnrollment.User.id,
      courseEnrollment.Course.id,
      bucket,
      false,
    );

    // Update the course enrollment record with the certificate URL
    await updateCourseEnrollmentRecord(
      courseEnrollment.User.id,
      courseEnrollment.Course.id,
      "achievementCertificateURL",
      path
    );
  } catch (error) {
    throw error;
  }
};

File: functions/callPythonFunction/pythonFunctions/generate_Achievement_Certificate.py
import asyncio
import requests
from graphqlclient import GraphQLClient
import logging

# Importieren Sie Ihre eigenen Hilfsfunktionen, ähnlich den `certificateUtils` in Ihrem Node.js-Code
from certificate_utils import get_storage_bucket_url, save_certificate_to_bucket, update_course_enrollment_record

logging.basicConfig(level=logging.DEBUG)

async def generate_achievement_certificate(course_enrollment, bucket):
    try:
        # Datenextraktion und Vorbereitung
        record_type = course_enrollment["User"]["AchievementRecordAuthors"][0]["AchievementRecord"]["AchievementOption"]["recordType"]
        record_title = course_enrollment["User"]["AchievementRecordAuthors"][0]["AchievementRecord"]["AchievementOption"]["title"]
        online_courses = record_title if record_type == "ONLINE_COURSE" else ""
        practical_project = record_title if record_type == "DOCUMENTATION" else ""

        # Zertifikatstext vorbereiten
        certificate_text = "\n".join([
            f"- {line.strip()}" for line in (course_enrollment["Course"]["learningGoals"] or "").split("\n") if line.strip() != ""
        ])

        # Zertifikatvorlage URL abrufen
        certificate_template_url = await get_storage_bucket_url(course_enrollment["Course"]["Program"]["achievementCertificateTemplateURL"], bucket)
        logging.debug(f"Certificate template URL retrieved: {certificate_template_url}")

        # Zertifikatsdaten konstruieren
        certificate_data = {
            "template": certificate_template_url,
            "full_name": f"{course_enrollment['User']['firstName']} {course_enrollment['User']['lastName']}",
            "semester": course_enrollment["Course"]["Program"]["title"],
            "course_name": course_enrollment["Course"]["title"],
            "ects": course_enrollment["Course"]["ects"],
            "practical_project": practical_project,
            "online_courses": online_courses,
            "certificate_text": certificate_text,
        }
        logging.debug(f"Certificate data: {certificate_data}")

        # Anfrage an den Zertifikatserstellungsdienst senden
        url = "https://edu-old.opencampus.sh/create_certificate_rest"
        response = requests.post(url, json=certificate_data)
        generated_certificate = response.json()

        # Generiertes Zertifikat speichern
        path = await save_certificate_to_bucket(
            generated_certificate,
            'achievement',
            course_enrollment['User']['id'],
            course_enrollment['Course']['id'],
            bucket,
            False
        )

        # Kursanmeldungsdatensatz aktualisieren
        await update_course_enrollment_record(
            course_enrollment['User']['id'],
            course_enrollment['Course']['id'],
            "achievementCertificateURL",
            path
        )
    except Exception as error:
        logging.error(f"Error generating certificate: {error}")
        raise

# Beispielhafter Aufruf der asynchronen Funktion
# asyncio.run(generate_achievement_certificate(course_enrollment, bucket))
